codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddSigv4aPlugin.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package software.amazon.smithy.aws.typescript.codegen;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+import software.amazon.smithy.codegen.core.SymbolProvider;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.knowledge.ServiceIndex;
+import software.amazon.smithy.model.knowledge.TopDownIndex;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.model.traits.Trait;
+import software.amazon.smithy.typescript.codegen.LanguageTarget;
+import software.amazon.smithy.typescript.codegen.TypeScriptSettings;
+import software.amazon.smithy.typescript.codegen.TypeScriptWriter;
+import software.amazon.smithy.typescript.codegen.integration.TypeScriptIntegration;
+import software.amazon.smithy.utils.MapUtils;
+import software.amazon.smithy.utils.SmithyInternalApi;
+
+/**
+ * Detects when to add sigv4a signer.
+ */
+@SmithyInternalApi
+public final class AddSigv4aPlugin implements TypeScriptIntegration {
+    public Map<String, Consumer<TypeScriptWriter>> getRuntimeConfigWriters(
+        TypeScriptSettings settings,
+        Model model,
+        SymbolProvider symbolProvider,
+        LanguageTarget target
+    ) {
+        boolean useSigv4aCapableSigner = false;
+
+        ServiceIndex serviceIndex = ServiceIndex.of(model);
+        Set<ServiceShape> services = model.getServiceShapes();
+        for (ServiceShape service : services) {
+            Map<ShapeId, Trait> authSchemes = serviceIndex.getAuthSchemes(service);
+            useSigv4aCapableSigner = useSigv4aCapableSigner || usesSigv4a(authSchemes);
+            if (useSigv4aCapableSigner) {
+                break;
+            }
+
+            TopDownIndex topDownIndex = TopDownIndex.of(model);
+            for (OperationShape operationShape : topDownIndex.getContainedOperations(service)) {
+                useSigv4aCapableSigner = usesSigv4a(operationShape.getAllTraits());
+                if (useSigv4aCapableSigner) {
+                    break;
+                }
+            }
+        }
```

Comment:
nit: I think a better way of checking this is:

1. If the service auth schemes don't include `sigv4a`, return empty
2. If all operation auth schemes don't have `sigv4a`, return empty
3. return `signerConstructor` with `SignatureV4MultiRegion`

This is because operation auth schemes are equal to or a subset of the service auth schemes.